spectec/spec/7-module.watsup

```diff
@@ -0,0 +1,134 @@
+;; Allocation
+
+
+def $funcs(externval*) : funcaddr*
+def $globals(externval*) : globaladdr*
+def $tables(externval*) : tableaddr*
+def $mems(externval*) : memaddr*
+
+def $funcs(externval*) = [fa|FUNC fa <- externval]
+def $globals(externval*) = [ga|GLOBAL ga <- externval]
+def $tables(externval*) = [ta|TABLE ta <- externval]
+def $mems(externval*) = [ma|MEM ma <- externval]
+
+
+def $instexport(funcaddr*, globaladdr*, tableaddr*, memaddr*, export) : exportinst
+def $instexport(fa*, ga*, ta*, ma*, EXPORT name (FUNC x)) = { NAME name, VALUE (FUNC fa*[x]) }
+def $instexport(fa*, ga*, ta*, ma*, EXPORT name (GLOBAL x)) = { NAME name, VALUE (GLOBAL ga*[x]) }
+def $instexport(fa*, ga*, ta*, ma*, EXPORT name (TABLE x)) = { NAME name, VALUE (TABLE ta*[x]) }
+def $instexport(fa*, ga*, ta*, ma*, EXPORT name (MEM x)) = { NAME name, VALUE (MEM ma*[x]) }
+
+
+def $allocfunc(moduleinst, func) : funcinst
+def $allocfunc(m, func) = { MODULE m, CODE func }
+
+def $allocglobal(globaltype, val) : globalinst
+def $allocglobal(globaltype, val) = { TYPE globaltype, VALUE val }
+
+def $alloctable(tabletype) : tableinst
+def $alloctable(`[i .. j] rt) = { TYPE (`[i .. j] rt), ELEM (REF.NULL rt)^i }
+
+def $allocmem(memtype) : meminst
+def $allocmem(`[i .. j] I8) = { TYPE (`[i .. j] I8), DATA 0^(i * 64 * $Ki()) }
+
+def $allocelem(reftype, ref*) : eleminst
+def $allocelem(rt, ref*) = { TYPE rt, ELEM ref* }
+
+def $allocdata(byte*) : datainst
+def $allocdata(byte*) = { DATA byte* }
+
+
+def $allocmodule(store, module, externval*, val*, (ref*)*) : (store, moduleinst)
+def $allocmodule(s, module, externval*, val*, (ref*)*) = (s_6, m)
+  -- if module =
+    MODULE
+      import*
+      func*
+      (GLOBAL globaltype expr)*
+      (TABLE tabletype)*
+      (MEMORY memtype)*
+      (ELEM rt expr* elemmode?)*
+      (DATA byte* datamode?)*
+      start?
+      export*
+  -- if fa_ex* = $funcs(externval*)
+  -- if ga_ex* = $globals(externval*)
+  -- if ta_ex* = $tables(externval*)
+  -- if ma_ex = $mems(externval*)
+  -- if fa* = |s.FUNC| ... $(|s.FUNC| + |func*| - 1)
+  -- if ga* = |s.GLOBAL| ... $(|s.GLOBAL| + |global*| - 1)
+  -- if ta* = |s.TABLE| ... $(|s.TABLE| + |table*| - 1)
+  -- if ma* = |s.MEM| ... $(|s.MEM| + |mem*| - 1)
+  -- if ea* = |s.ELEM| ... $(|s.ELEM| + |elem*| - 1)
+  -- if da* = |s.DATA| ... $(|s.DATA| + |data*| - 1)
+  -- if xi* = $instexport(fa_ex* fa*, ga_ex* ga*, ta_ex* ta*, ma_ex* ma*, export)*
+  -- if m = {
+      FUNC fa_ex* fa*,
+      GLOBAL ga_ex* ga*,
+      TABLE ta_ex* ta*,
+      MEM ma_ex* ma*,
+      ELEM ea*,
+      DATA da*,
+      EXPORT xi*
+    }
+  -- if s_1 = s[.FUNC =.. $allocfunc(m, func)*]
+  -- if s_2 = s_1[.GLOBAL =.. $allocglobal(globaltype, val)*]
+  -- if s_3 = s_2[.TABLE =.. $alloctable(tabletype)*]
+  -- if s_4 = s_3[.MEM =.. $allocmem(memtype)*]
+  -- if s_5 = s_4[.ELEM =.. $allocelem(rt, ref*)*]
+  -- if s_6 = s_5[.DATA =.. $allocdata(byte*)*]
```

Comment:
Now, we don't need allocx* liftings.
But we can use allocx* lifting, if you want.
```
def $allocfunc(store, moduleinst, func) : store
def $allocfunc(s, m, func) = s.[.FUNC =.. fi]
  -- if fi = { MODULE m, CODE func }

def $allocfuncs(store, moduleinst, func*) : store
def $allocfuncs(s, m, epsilon) = s
def $allocfuncs(s, m, func func'*) = $allocfuncs($allocfunc(s, m, func), m, func'*)

def $allocglobal(store, global, val) : store
def $allocglobal(s, GLOBAL globaltype instr*, val) = s[.GLOBAL =.. gi]
  -- if gi = { TYPE globaltype, VALUE val }

def $allocglobals(store, global*, val*) : store
def $allocglobals(s, epsilon, epsilon) = s
def $allocglobals(s, global global'*, val val'*) =
  $allocglobals($allocglobal(s, global, val), global'*, val'*)

def $alloctable(store, table) : store
def $alloctable(s, TABLE tabletype) = s[.TABLE =.. ti]
  -- if tabletype = `[i .. j] rt
  -- if ti = { TYPE tabletype, ELEM (REF.NULL rt)^i }

def $alloctables(store, table*) : store
def $alloctables(s, epsilon) = s
def $alloctables(s, table table'*) = $alloctables($alloctable(s, table), table'*)

def $allocmem(store, mem) : store
def $allocmem(s, MEMORY memtype) = s[.MEM =.. mi]
  -- if memtype = `[i .. j] I8
  -- if mi = { TYPE memtype, DATA 0^(i * 64 * $Ki()) }

def $allocmem*(store, mem*) : store
def $allocmem*(s, epsilon) = s
def $allocmem*(s, mem mem'*) = $allocmem*($allocmem(s, mem), mem'*)

def $allocelem(store, elem, ref*) : store
def $allocelem(s, ELEM rt (instr*)* elemode?, ref*) = s[.ELEM =.. ei]
  -- if ei = { TYPE rt, ELEM ref* }

def $allocelem*(store, elem*, (ref*)*) : store
def $allocelem*(s, epsilon, epsilon) = s
def $allocelem*(s, elem elem'*, ref* (ref'*)*) =
  $allocelem*($allocelem(s, elem, ref*), elem'*, (ref'*)*)

def $allocdata(store, data) : store
def $allocdata(s, DATA byte* datamode?) = s[.DATA =.. di]
  -- if di = { DATA byte* }

def $allocdata*(store, data*) : store
def $allocdata*(s, epsilon) = s
def $allocdata*(s, data data'*) = $allocdata*($allocdata(s, data), data'*)


def $allocmodule(store, module, externval*, val*, (ref*)*) : (store, moduleinst)
def $allocmodule(s, module, externval*, val*, (ref*)*) = (s_6, m)
  -- if module =
    MODULE
      import*
      func*
      (GLOBAL globaltype expr)*
      (TABLE tabletype)*
      (MEMORY memtype)*
      (ELEM rt expr* elemmode?)*
      (DATA byte* datamode?)*
      start?
      export*
  -- if fa_ex* = $funcs(externval*)
  -- if ga_ex* = $globals(externval*)
  -- if ta_ex* = $tables(externval*)
  -- if ma_ex = $mems(externval*)
  -- if fa* = |s.FUNC| ... $(|s.FUNC| + |func*| - 1)
  -- if ga* = |s.GLOBAL| ... $(|s.GLOBAL| + |global*| - 1)
  -- if ta* = |s.TABLE| ... $(|s.TABLE| + |table*| - 1)
  -- if ma* = |s.MEM| ... $(|s.MEM| + |mem*| - 1)
  -- if ea* = |s.ELEM| ... $(|s.ELEM| + |elem*| - 1)
  -- if da* = |s.DATA| ... $(|s.DATA| + |data*| - 1)
  -- if xi* = $instexport(fa_ex* fa*, ga_ex* ga*, ta_ex* ta*, ma_ex* ma*, export)*
  -- if m = {
      FUNC fa_ex* fa*,
      GLOBAL ga_ex* ga*,
      TABLE ta_ex* ta*,
      MEM ma_ex* ma*,
      ELEM ea*,
      DATA da*,
      EXPORT xi*
    }
  -- if s_1 = $allocfunc*(s, m, func*)
  -- if s_2 = $allocglobal*(s_1, globaltype*, val*)
  -- if s_3 = $alloctable*(s_2, tabletype*)
  -- if s_4 = $allocmem*(s_3, memtype*)
  -- if s_5 = $allocelem*(s_4, rt*, (ref*)*)
  -- if s_6 = $allocdata*(s_5, (byte*)*)
```

Reply:
The alloc functions are meant as an abstraction that encapsulates allocation (and they are used in other places, specifically, the embedding interface in the appendix). The formulation you suggest, while cute, would be problematic in that sense, because then these functions no longer actually allocate.

So I'd go with the liftings for now. (You probably need to name them without the star and make that a show-hint.)

Mid-term I'd like to revamp the module representation into a linear list of declaration that is processed sequentially. Then this mess should go away.

Reply:
I have a few questions related to the comments you made:
1. To avoid mutation, we decided to precompute the addresses before allocation. As a result, the alloc functions no longer return address. Is it acceptable to return store only?
2. In the spec, the allocx lifting is defined for arbitrary object kinds. However, DSL doesn't allow it, so I define a lifting function for each of the alloc functions. Is this approach acceptable?
3. You mentioned that it is necessary to name the lifting functions without a star. If we remove the star, the name of lifting function would be the same as the name of allocation function. To ensure distinct names, would it be appropriate to use the name `allocfuncs` for the lifting function?
4. I didn't fully understand the sentence: "Mid-term I'd like to revamp the module representation into a linear list of declaration that is processed sequentially." Could you provide further clarification on its meaning?

Reply:
Re 1: Well, if you rephrase "precompute" as "predict", then no change to the allocation functions is necessary, I believe.

Re 2: Yes, defining a function for each type is fine for the DSL. We can probably still tweak it to produce a single template in the Latex output.

Re 3: `allocfuncs` sounds good to me.

Re 4: With some of the future extensions, module checking and instantiation becomes more sequential, with ordered dependencies between declarations. Also, there are ideas for allowing multiple appearances of section types, so that the order of different sorts of declarations can be mixed. So I would like to refactor modules to be more like a sequence of different declarations that is processed in order, which would also eliminate the specific iterations. I still have to work out the details, though, in particular wrt instantiation.

Reply:
Thank you for your kind reply!